sql/core/src/main/scala/org/apache/spark/sql/execution/SparkQl.scala

```diff
@@ -62,6 +66,458 @@ private[sql] class SparkQl(conf: ParserConf = SimpleParserConf()) extends Cataly
         val tableIdent = extractTableIdent(nameParts)
         RefreshTable(tableIdent)
 
+      case Token("TOK_CREATEDATABASE", Token(databaseName, Nil) :: createDatabaseArgs) =>
```

Comment:
The main problem is LogicalPlan parsing, which we can split up in command/ddl/query parsing. We could use partial functions (or something like that) to implement the different parts of the parsing logic.

For instance:

    abstract class BaseParser(val conf: ParserConf) extends ParserInterface {
      val planParsers: Seq[PlanParser]

      lazy val planParser = planParsers.reduce(_.orElse(_))

      def nodeToPlan(node: ASTNode): LogicalPlan = {
        planParser.applyOrElse(node, throw new NotImplementedError(node.text))
      }
    }

    abstract class PlanParser extends PartialFunction[ASTNode, LogicalPlan]

    case class ExplainCommandParser(base: BaseParser) extends PlanParser {
      val catalystQl = new CatalystQl
      import catalystQl.Token
      
      override def isDefinedAt(node: ASTNode): Boolean = node.text == "TOK_EXPLAIN"

      override def apply(v1: ASTNode): LogicalPlan = v1.children match {
        case (crtTbl @ Token("TOK_CREATETABLE" | "TOK_QUERY", _)) :: rest =>
          val extended = rest.exists(_.text.toUpperCase == "EXTENDED")
          ExplainCommand(base.nodeToPlan(crtTbl), extended)
      }
    }

    class SomeParser(conf: ParserConf) extends BaseParser {
      val planParsers: Seq[PlanParser] = Seq(
        ExplainCommandParser(this))
    }